infra/compile.py

```diff
@@ -300,6 +299,10 @@ def load_outline(module):
 def compile_method(base, name, args, ctx):
     base_js = compile_expr(base, ctx)
     args_js = [compile_expr(arg, ctx) for arg in args]
+    needs_call = False
+    if base_js[0].isupper():
+        base_js = base_js + ".prototype"
+        needs_call = True
```

Comment:
What's the Python code that needs this? It seems reasonable but I'm not sure what it's for.

Reply:
It's for chapter 16, which instead of calling `super().layout()` for `EmbedLayout` subclasses, calls `EmbedLayout.layout`. This code detects that pattern and uses the `call` syntax which is equivalent to it, and allows passing `this`.

The reason lab16 uses that pattern in Python is because it's using the patching mechanism, which messes up inheritance, and using `super()` therefore throws an exception saying you can't do that if one class is not inheriting from another.